Add RKE2 as an additional kubernetes distribution

mikeywuu opened the pull request:
This PR addresses the implementation of changes relevant to enable users to deploy kubernetes clusters with RKE2 instead of k3s. For some details, please see the discussion around this #1132.
I have tried to be as DRY as possible, but some things, especially resources which could produce unneccessary changes, are separated.

So far I am only using Cilium as CNI, so all others are still open to be adjusted and properly tested. I moved the docs for Backup & Restore in to another file in the docs folder, feel free to undo that in case it should stay in the global README.

To switch between a k3s and RKE2 cluster, I have added a variable `kubernetes_distribution_type`. This is the variable where a lot of the switching in the background happens. Additionally, there is a new variable `install_rke2_version` to manage the version to install. It defaults to the current stable version.

I have added some TODO comments to the code where one would need to put some attention to. Apart from that, I have these open points:

- [ ] Other CNI's than cilium
- [ ] I have not yet managed to make the needed adjustments to make selinux work. I added some things from the official repo (https://github.com/rancher/rke2-selinux/tree/master) but I'm not experienced enough to fix that so I disabled it for my cluster. Apart from that the output while the servers are bootstrapped in the `terraform apply` run are complaining about `the installed k3s-selinux-1.6-1.sle.noarch conflicts with 'rke2-selinux' provided by the to be installed rke2-selinux-0.18-1.sle.noarch`. Not sure how to fix that. Also I stumbled across the reboot issue that is currently fixed by running a specific command. Selinux partially worked for me after rebooting nodes manually, but that wasn't acceptable enough for me
- [ ] Autoscaler nodes can be deployed but seem to have some networking issues. I cannot see logs or exec into the containers. Also metrics are not displayed in e.g. tools like `k9s`
- [ ] The kubelet reservations seem a bit high when using `cax11` instances. In my case the cluster-autoscaler couldn't be scheduled because the control planes had to much memory requests already present. My workaround for now was the `kubelet_args    = []` line in the nodepools.

<details>
  <summary>I deploy my cluster with this kubeconfig</summary>
  
```
module "kube_hetzner" {

  source                        = "/workspace/repos/infrastructure/terraform/modules/terraform-hcloud-kube-hetzner"
  providers                     = { hcloud = hcloud }
  hcloud_token                  = var.hcloud_token
  hcloud_ssh_key_id             = data.hcloud_ssh_key.infra.id
  ssh_public_key                = file("~/.ssh/id_rsa_hetzner_infra.pub")
  ssh_private_key               = file("~/.ssh/id_rsa_hetzner_infra")
  ssh_additional_public_keys    = [file("~/.ssh/id_rsa_ansible.pub"), ]
  existing_network_id           = [data.hcloud_network.network1.id]
  network_ipv4_cidr             = "10.128.0.0/16"
  cluster_name                  = "rke2-test-blue"
  hetzner_ccm_use_helm          = true
  kubernetes_distribution_type  = "rke2"
  k3s_token                     = local.k3s_token
  use_cluster_name_in_node_name = false
  install_rke2_version          = "v1.32.5+rke2r1"
  kured_options = {
    "log-format" : "json",
    "start-time" : "8pm",
  }
  disable_selinux           = true
  automatically_upgrade_k3s = true
  automatically_upgrade_os  = true
  enable_cert_manager       = false
  disable_kube_proxy        = true
  cni_plugin                = "cilium"
  cilium_routing_mode       = "native"
  cilium_hubble_enabled     = true
  cilium_values             = <<EOT
envoy:
  enabled: false
MTU: 1450
bpf:
  masquerade: true
endpointRoutes:
  enabled: true
hubble:
  metrics:
    enabled:
    - dns
    - drop
    - tcp
    - flow
    - port-distribution
    - icmp
    - httpV2:exemplars=true;labelsContext=source_ip,source_namespace,source_workload,destination_ip,destination_namespace,destination_workload,traffic_direction
    enableOpenMetrics: true
  relay:
    enabled: true
  ui:
    enabled: true
installNoConntrackIptablesRules: true
ipam:
  mode: kubernetes
ipv4NativeRoutingCIDR: 10.42.0.0/16
k8s:
  requireIPv4PodCIDR: true
k8sServiceHost: 127.0.0.1
k8sServicePort: "6443"
kubeProxyReplacement: true
kubeProxyReplacementHealthzBindAddr: 0.0.0.0:10256
loadBalancer:
  acceleration: native
routingMode: native
encryption:
  enabled: true
  type: wireguard
  nodeEncryption: true
EOT
  ingress_controller        = "none"
  restrict_outbound_traffic = false
  firewall_kube_api_source  = local.firewall_kube_api_source
  firewall_ssh_source       = local.firewall_ssh_source
  control_planes_custom_config = {
    etcd-expose-metrics         = true,
    kube-controller-manager-arg = "bind-address=0.0.0.0",
    kube-scheduler-arg          = "bind-address=0.0.0.0",
  }
  etcd_s3_backup = {
    etcd-s3-endpoint   = "myvalue"
    etcd-s3-access-key = "myvalue"
    etcd-s3-secret-key = "myvalue"
    etcd-s3-bucket     = "myvalue"
    etcd-s3-folder     = "myvalue"
    etcd-s3-region     = "myvalue"
  }
  control_plane_nodepools = [
    {
      name            = "rke2-control-test-1",
      server_type     = "cax11",
      location        = "fsn1",
      labels          = [],
      taints          = [],
      count           = 1,
      selinux         = false
      placement_group = "rke2-control-test-1"
      kubelet_args    = []
    },
    {
      name            = "rke2-control-test-2",
      server_type     = "cax11",
      location        = "fsn1",
      labels          = [],
      taints          = [],
      count           = 1,
      selinux         = false,
      placement_group = "rke2-control-test-1"
      kubelet_args    = []
    },
    {
      name            = "rke2-control-test-3",
      server_type     = "cax11",
      location        = "fsn1",
      labels          = [],
      taints          = [],
      count           = 1,
      selinux         = false,
      placement_group = "rke2-control-test-1"
      kubelet_args    = []
    },
  ]
  agent_nodepools = [
    # {
    #   name            = "rke2-worker-test",
    #   server_type     = "cax11",
    #   location        = "fsn1",
    #   labels          = [],
    #   taints          = [],
    #   placement_group = "rke2-worker-test"
    #   selinux         = false
    #   nodes = {
    #     "1" : {
    #       append_index_to_node_name = true,
    #     },
    #     # "2" : {
    #     #   append_index_to_node_name = true,
    #     # },
    #   }
    # },
  ]
  autoscaler_nodepools = [
    {
      name        = "rke2-worker-r"
      server_type = "cax11"
      location    = "fsn1"
      min_nodes   = 1
      max_nodes   = 4
      selinux     = false
      labels = {
        "node.kubernetes.io/role" : "regular-workloads"
      }
    }
  ]
}
```
</details>

Looking forward to complete this feature together with you guys :)

mikeywuu committed:
feat: refactor to allow usage of RKE2

A RKE2 + cilium based cluster can be deployed

mikeywuu committed:
feat: refactor to allow usage of RKE2

A RKE2 + cilium based cluster can be deployed

mikeywuu committed:
wip

mikeywuu committed:
wip / debug

mikeywuu committed:
remove comments to avoid resource diffs

mikeywuu committed:
make sure rke2 config is properly created

mikeywuu committed:
disable selinux wip

mikeywuu committed:
wip

mikeywuu committed:
Merge branch 'master' into add-rke2-wip

# Conflicts:
#	docs/terraform.md

mikeywuu committed:
configure upgrade plans for upgrading rke2

mikeywuu committed:
adjust selinux configuration to work with RKE2

mikeywuu committed:
Merge branch 'refs/heads/master' into add-rke2-wip

mikeywuu committed:
merge and apply the latest changes also to RKE2 resources

mikeywuu committed:
add docs for backup and restore of rke2

mikeywuu committed:
add labels to autoscaler nodes; align latest master changes

mikeywuu committed:
make sure that kubelet-arg matches on first and all following control nodes to have the same resource reservations

mikeywuu committed:
Merge branch 'master' into add-rke2-wip

# Conflicts:
#	.gitignore
#	autoscaler-agents.tf
#	locals.tf

mikeywuu committed:
align upstream changes in relevant RKE2 parts

mikeywuu committed:
align upstream changes in relevant RKE2 parts

mikeywuu commented:
@mysticaltech do you have an update on when we could take a look at this one? Then I would make sure to sync recent changes.
Also, I was wondering if we could talk about the future OS to be introduced in #1628. It would be great to have RKE2 running on the new OS from the beginning to avoid any possible migrations later. Also, I'm not yet sure how to implement a switch for  k3s / RKE2 in the packer template, since there are few selinux packages specific to k3s. Installing the RKE2 libraries on top has produced some errors in my testing.

mysticaltech commented:
@mikeywuu Appreciate this work very much. I'm going to merge the LeapMicro PR first then, I will properly review this. As for the SELinux packages in packer, you are right that this is problematic. We will probably just need to move the installation outside of packer, even if it adds one reboot to the node. There are places, I will take a closer look soon.

mikeywuu committed:
fix kured release yaml

mikeywuu committed:
make authentication-config arg for kube-apiserver dynamic to support both k3s and rke2

mikeywuu commented:
@mysticaltech some time has passed and my fork has fallen quite far behind. Since the OS change doesn’t seem to be coming anytime soon, I think it makes sense to plan the next steps now.

At this point, it might actually be easier to re-implement things from scratch so everything aligns with the current version.

If you could take a quick look at the current changes and share some initial feedback, that would help a lot. From my side, a re-implementation would be totally fine.

mysticaltech committed:
fix: support authentication config path for rke2

mysticaltech committed:
fix: align rke2 auth config path handling and service restarts

Use distribution-specific authentication_config path and service names without introducing undeclared audit variables in this draft branch.

mysticaltech committed:
Merge staging into add-rke2-wip and resolve conflicts

mysticaltech committed:
Merge staging into add-rke2-wip and resolve kustomization dependency conflict
